include/graph_contracts.hpp

```diff
@@ -0,0 +1,444 @@
+            }
+        }
+        ~wrapper() override = default;
+        [[nodiscard]] constexpr supported_type pmt_type() const noexcept override { return _value.pmt_type(); }
+        [[nodiscard]] constexpr port_type_t type() const noexcept override { return T::type(); }
+        [[nodiscard]] constexpr port_direction_t direction() const noexcept override { return T::direction(); }
+        [[nodiscard]] constexpr std::string_view name() const noexcept override {  return _value.name(); }
+        [[nodiscard]] connection_result_t resize_buffer(std::size_t min_size) noexcept override {  return _value.resize_buffer(min_size); }
+        [[nodiscard]] connection_result_t disconnect() noexcept override {  return _value.disconnect(); }
+        [[nodiscard]] connection_result_t connect(dyn_port& dst_port) noexcept override {
+            auto src_buffer = _value.writer_handler_internal();
+            return dst_port.update_reader_internal(src_buffer) ? connection_result_t::SUCCESS: connection_result_t::FAILED;
+        }
+    };
+
+    bool update_reader_internal(void* buffer_other)  noexcept { return _accessor->update_reader_internal(buffer_other); }
+
+public:
+    using value_type = void;
+    template<Port T>
+    explicit constexpr dyn_port(T &arg) noexcept : _accessor{ std::make_unique<wrapper<T>>(arg) } {}
+
+    [[nodiscard]] supported_type pmt_type() const noexcept { return _accessor->pmt_type(); }
+    [[nodiscard]] port_type_t type() const noexcept { return _accessor->type(); }
+    [[nodiscard]] port_direction_t direction() const noexcept { return _accessor->direction(); }
+    [[nodiscard]] std::string_view name() const noexcept { return _accessor->name(); }
+    [[nodiscard]] connection_result_t resize_buffer(std::size_t min_size) {
+        if (direction() == port_direction_t::OUTPUT) {
+            return _accessor->resize_buffer(min_size);
+        }
+        return connection_result_t::FAILED;
+    }
+    [[nodiscard]] connection_result_t disconnect() noexcept {  return _accessor->disconnect(); }
+    [[nodiscard]] connection_result_t connect(dyn_port& dst_port) noexcept { return _accessor->connect(dst_port); }
+};
+static_assert(Port<dyn_port>);
+
+class block { // TODO: test block for development purposes to be merged/replaced with impl in graph.hpp
+    using setting_map = std::map<std::string, int, std::less<>>;
+
+    std::string _name;
+    std::vector<dyn_port> _input_ports{};
+    std::vector<dyn_port> _output_ports{};
+    setting_map _exec_metrics{}; //  →  std::map<string, pmt> → fair scheduling, 'int' stand-in for pmtv
+
+public:
+    explicit block(std::string_view name) : _name(name) {};
+
+    [[nodiscard]] std::string_view name() const noexcept { return _name; }
+    void work() noexcept { /* to be implemented in derived classes */ }
+    [[nodiscard]] std::vector<dyn_port>const & input_ports() const noexcept { return _input_ports; }
```

Comment:
Not a fan of giving access to the underlying structure to user.  IMO, this should return std::span, and all modifications should go through proper API, not through vector's API.

Reply:
changed and shielded the access via explicit `add_port(..)` and `port<INPUT(...)` accessor that can be queried via string and/or id (N.B. should be deprecated for user-api because indices are error prone). Also having no removal operation should make it safer in general w.r.t. initialisation and life-time management.